### PR TITLE
[PR] embedding layer

### LINKED ISSUE #11: [FEATURE] embedding layer
## Feature Description

해당 Issue는 transformer의 embedding layer를 구현합니다.

## Motivation and Context

embedding layer는 transformer 구조만이 아니라 다양한 nlp 모델에 입력으로 사용되는 기초적인 아키텍쳐입니다.
embedding layer는 단어를 밀집 표현으로 변환해 벡터로 표현하며, 초기화된 임베딩은 학습 과정을 통해 조정되어 단어들 사이의 의미론적 유사성을 반영하게 됩니다. transformer에서는 embedding layer 이후 positional encoding layer와 더해져서 encoder의 입력으로 사용됩니다. 
pytorch에서는 nn.Embedding()으로 간단하게 구현되는 부분을 직접 구현했습니다. 

## Proposed Solution

임베딩 층의 입력으로 사용하기 위해서 입력 시퀀스의 각 단어들은 모두 정수 인코딩이 되어있다고 가정했습니다.

- [x] Embedding 구조체 정의 : 임베딩 정보를 저장하기 위한 기본 구조체를 정의, 2차원 float32 slice 
- [x] NewEmbedding 함수 정의 : 임베딩 구조체를 생성하고 초기화하는 생성자(constructor) 역할을 하는 함수, 임베딩할 단어의 개수와 임베딩 벡터의 차원을 입력으로 받습니다.
- [x] 임의의 값으로 벡터 초기화 : rand.Float32()함수를 사용해 0과 1사이 임의의 값으로 초기화 됨.

## Alternatives Considered

nn.Embedding() 함수에는 padding 또한 고려하는 부분이 있으나 해당 부분은 optional한 부분이기 때문에 추후 개선하도록 하겠습니다.

## Additional Information

추후 develop 할 부분
- [ ] positional encoding part와 연결되어 encoding layer로 들어가는 코드 작성하기

### PR DESCRIPTION
**Pull request by ohbigkite**
---
name: Pull Request
about: Create a pull request to contribute to the project
title: '[PR] Brief description of the changes'
labels: 'Type/Feature, Progress/In Progress'
assignees: ''

---

## Description

transformer의 encoder의 입력으로 사용되는 embedding layer를 구현했습니다. 임베딩값은 임베딩 단어 개수 * 임베딩 차원 형태의 룩업테이블 형식으로 생성됩니다.

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

embedding_test.go에서 n_input_vocab, d_model을 입력받아 n_input_vocab * d_model 형태의 2차원 배열이 생성되어 랜덤값으로 초기화되는지 확인했습니다.

## Checklist:
Please check and confirm the following 

- [ ] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

## Related Issue
resolve #11 

## Motivation and Context
추후 positional encoding part와 연결되어 transformer의 encoder layer의 입력으로 사용될 수 있습니다.

### CODE REVIEW

File: embedding.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package embedding
```
Comment (kwanok) [inline embedding.go:1]: Go에서는 package 이름을 디렉토리(폴더)의 이름으로 정하는 관례가 있습니다. 
좀 더 확실하게 얘기하면 패키지 이름을 디렉토리 이름으로 하는 것이죠.
다만 root 경로의 패키지는 예외적으로 main이라고 합니다.

따라서 해당 패키지는 다음과 같이 embedding 디렉토리 안으로 옮기는 것이 좋아보입니다.
```
goransformer/
├── embedding/
│   ├── embedding.go
│   └── embedding_test.go
├── go.mod
└── main.go
```

더 자세한 Package 이름 규칙은 다음 링크에서 자세하게 확인할 수 있습니다.

- https://go.dev/blog/package-names

File: embedding.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+)
+
+// Embedding 구조체 정의
+type Embedding struct {
+	Embeddings [][]float32
+}
+
+// NewEmbedding 함수는 n_input_vocab와 d_model을 받아 초기화된 Embedding 구조체를 반환
+// n_input_vocab : 임베딩을 할 단어들의 개수. d_model : 임베딩 할 벡터의 차원
+func NewEmbedding(n_input_vocab, d_model int) *Embedding {
```
Comment (kwanok) [inline embedding.go:14]: 함수로 넘겨주는 인자는 이런 부분이 크게 상관이 없지만, Convention이니 설명을 해보겠습니다.

Go 에서는 `snake_case`보다는 `MixedCaps` 또는 `mixedCaps` 를 사용하는 것이 좋습니다. 

이유는 Go 에서 맨 앞 글자가 대문자이냐 소문자이냐에 따라서 변수의 public, private을 결정하기 때문입니다.

`myEmbedding` - private, 이 코드에서 embedding에서만 사용 가능함. **다른 패키지에서 사용 불가**
`MyEmbedding` - public, 다른 패키지에서 embedding을 import 하면 접근이 가능함.

해당 코드는 다음과 같이 변경할 수 있습니다.

```go
func NewEmbedding(nInputVocab, dModel int) *Embedding {
    ...
}
```

**Review by kwanok — CHANGES_REQUESTED** [review]

File: embedding.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
+package embedding
+
+import (
+	"math/rand"
+)
+
+// Embedding 구조체 정의
+type Embedding struct {
+	Embeddings [][]float32
```
Comment (kwanok) [inline embedding.go:9]: 해당 부분이 float32라서 다른 분들과 구현한 것과 Conflict가 생길 것 같아서요

우선 float64로 적용하는게 어떨까요? 다른 분들 의견 여쭤봅니다.

@maxseats @alwaysday4u
